### PR TITLE
Extracted dotnet process calls to callbacks

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
https://github.com/dotnet/templating/issues/4518; dotnet/templating part

### Solution
- Extracted dotnet process calls to `NewCommandCallbacks`
- Implemented them via `dotnet` process run for `dotnet-new3`.
- Improved post action tests

### Checks:
- [ ] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: test/dotnet-new3.UnitTests/Approvals/PostActionTests.RunScript_Basic_Approval.OSX.verified.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿The template "TestAssets.PostActions.RunScript.Basic" was created successfully.
+
+Processing post-creation actions...
+
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/Approvals/PostActionTests.RunScript_Basic_Approval.OSX.verified.txt:4]: Need to investigate why empty line is appearing on Unix system here. Not related to this PR, may be done in separate issue if meets the bar.

File: src/dotnet-new3/New3CommandFactory.cs [hunk 2]
```diff
@@ -27,12 +27,41 @@ internal static Command Create()
...
             newCommand.AddGlobalOption(_debugDisableBuiltInTemplatesOption);
             newCommand.AddCommand(new CompleteCommand());
             return newCommand;
         }
+
+        private static bool ExecuteDotnetCommand(Dotnet command)
+        {
+            command.CaptureStdOut();
+            command.CaptureStdErr();
+            Console.WriteLine($"Running '{command.Command}':");
```
Comment (JanKrivanek) [inline src/dotnet-new3/New3CommandFactory.cs:48]: We should avoid writing directly to `Console`

File: src/dotnet-new3/New3CommandFactory.cs [hunk 3]
```diff
@@ -27,12 +27,41 @@ internal static Command Create()
...
+            var result = command.Execute();
+            Console.WriteLine(result.ExitCode == 0 ? "Command succeeded." : "Command failed.");
+            WriteCommandOutput(result);
+            return result.ExitCode == 0;
+        }
+
+        /// <summary>
+        /// Writes formatted command output from <paramref name="process"/>.
+        /// </summary>
+        private static void WriteCommandOutput(Dotnet.Result process)
```
Comment (JanKrivanek) [inline src/dotnet-new3/New3CommandFactory.cs:58]: Why do we want all this output in `Console`?

File: src/Microsoft.TemplateEngine.Cli/NewCommandCallbacks.cs [hunk 4]
```diff
@@ -21,5 +21,31 @@ public sealed class NewCommandCallbacks
...
         /// </summary>
         public Func<string, bool>? RestoreProject { get; set; }
+
+        /// <summary>
+        /// Callback to be executed to add reference to a project.
+        /// Parameters: <br/>
+        /// - project name (string) <br/>
+        /// - arguments (IReadOnlyList&lt;string&gt;).
+        /// </summary>
+        public Func<string, IReadOnlyList<string>, bool>? AddProjectReference { get; set; }
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Cli/NewCommandCallbacks.cs:31]: Not introduced by this PR - but should we get rid of the public setters? It is being passed all over, but we likely do not want any swaping of the initial callbacks

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks good.
I'm just not sure if direct writing to console was just testing leftover or if it's intentiaonal (then I'm unsure about the intention)

File: src/dotnet-new3/New3CommandFactory.cs (hunk 2 above)
Comment (vlada-shubina) [reply src/dotnet-new3/New3CommandFactory.cs:48]: `dotnet-new3` doesn't have access to `Reporter`.

File: src/dotnet-new3/New3CommandFactory.cs (hunk 3 above)
Comment (vlada-shubina) [reply src/dotnet-new3/New3CommandFactory.cs:58]: `dotnet-new3` doesn't have access to `Reporter`.

File: src/Microsoft.TemplateEngine.Cli/NewCommandCallbacks.cs (hunk 4 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/NewCommandCallbacks.cs:31]: this is a valid concern, I'll remove them

Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/NewCommandCallbacks.cs]: changed to `init`

**Pull request merged**